File: Sources/MouseBridge/CommandLineInterface.swift
import ApplicationServices
import Foundation

@MainActor
enum CommandLineInterface {
    static func run(arguments: [String]) -> Int? {
        guard let command = arguments.first else { return nil }
        switch command {
        case "config": return runConfig(Array(arguments.dropFirst()))
        case "launch-at-login": return runLaunchAtLogin(Array(arguments.dropFirst()))
        case "diagnose":
            print("config=\(ConfigStore.shared.fileURL.path)")
            print("log=\(DiagnosticLog.shared.fileURL.path)")
            print("accessibility=\(AXIsProcessTrusted())")
            print("input-monitoring=\(CGPreflightListenEventAccess())")
            print("launch-at-login=\(LaunchAtLoginController.status.description)")
            return 0
        case "help", "--help", "-h":
            printHelp()
            return 0
        default: return nil
        }
    }

    private static func runLaunchAtLogin(_ arguments: [String]) -> Int {
        guard arguments.count == 1 else {
            return fail("用法：MouseBridge launch-at-login status|enable|disable", code: 2)
        }
        do {
            switch arguments[0] {
            case "status":
                print(LaunchAtLoginController.status.description)
            case "enable":
                try LaunchAtLoginController.setEnabled(true)
                print(LaunchAtLoginController.status.description)
            case "disable":
                try LaunchAtLoginController.setEnabled(false)
                print(LaunchAtLoginController.status.description)
            default:
                return fail("未知 launch-at-login 操作：\(arguments[0])", code: 2)
            }
            return 0
        } catch {
            return fail("登录启动设置失败：\(error.localizedDescription)")
        }
    }

    private static func runConfig(_ arguments: [String]) -> Int {
        guard let operation = arguments.first else {
            printHelp()
            return 2
        }
        switch operation {
        case "path":
            print(ConfigStore.shared.fileURL.path)
            return 0
        case "get":
            do {
                FileHandle.standardOutput.write(try ConfigStore.shared.config.prettyJSONData())
                FileHandle.standardOutput.write(Data("\n".utf8))
                return 0
            } catch {
                return fail(error.localizedDescription)
            }
        case "set":
            guard arguments.count == 3 else {
                return fail("用法：MouseBridge config set <key> <value>", code: 2)
            }
            var config = ConfigStore.shared.config
            let key = arguments[1]
            let value = arguments[2]
            switch key {
            case "middle": config.middleAction = value
            case "back": config.backAction = value
            case "forward": config.forwardAction = value
            case "reverse-vertical":
                guard let parsed = parseBool(value) else { return fail("布尔值应为 true/false", code: 2) }
                config.reverseVerticalScroll = parsed
            case "reverse-horizontal":
                guard let parsed = parseBool(value) else { return fail("布尔值应为 true/false", code: 2) }
                config.reverseHorizontalScroll = parsed
            case "scroll-lines":
                guard let parsed = Int(value), (0...20).contains(parsed) else { return fail("scroll-lines 应为 0–20", code: 2) }
                config.scrollLines = parsed
            case "dpi":
                guard let parsed = Int(value), (400...4000).contains(parsed) else { return fail("dpi 应为 400–4000", code: 2) }
                config.primaryDPI = parsed
            default:
                return fail("未知配置项：\(key)", code: 2)
            }
            guard [config.middleAction, config.backAction, config.forwardAction].allSatisfy(ShortcutExecutor.isValid) else {
                return fail("快捷键格式无效", code: 2)
            }
            do {
                try ConfigStore.shared.save(config)
                print("ok")
                return 0
            } catch {
                return fail(error.localizedDescription)
            }
        default:
            return fail("未知 config 操作：\(operation)", code: 2)
        }
    }

    private static func parseBool(_ value: String) -> Bool? {
        switch value.lowercased() {
        case "true", "1", "yes", "on": return true
        case "false", "0", "no", "off": return false
        default: return nil
        }
    }

    private static func fail(_ message: String, code: Int = 1) -> Int {
        FileHandle.standardError.write(Data("MouseBridge: \(message)\n".utf8))
        return code
    }

    private static func printHelp() {
        print("""
        MouseBridge commands:
          MouseBridge config path
          MouseBridge config get
          MouseBridge config set middle|back|forward <shortcut|none>
          MouseBridge config set reverse-vertical|reverse-horizontal <true|false>
          MouseBridge config set scroll-lines <0-20>
          MouseBridge config set dpi <400-4000>
          MouseBridge launch-at-login status|enable|disable
          MouseBridge diagnose
        """)
    }
}

File: platforms/windows/MouseBridge/CommandLineInterface.cs
// SPDX-License-Identifier: GPL-3.0-or-later

using System.Reflection;
using System.Runtime.InteropServices;
using System.Text.Json;

namespace MouseBridge.Windows;

internal static class CommandLineInterface
{
    private static readonly JsonSerializerOptions PrettyJson = new() { WriteIndented = true };

    public static int Run(string[] arguments)
    {
        if (arguments.Length == 0) return 2;
        try
        {
            return arguments[0] switch
            {
                "config" => RunConfig(arguments[1..]),
                "diagnose" when arguments.Length == 1 => Diagnose(),
                "help" or "--help" or "-h" when arguments.Length == 1 => Help(0),
                _ => Fail($"unknown command: {arguments[0]}", 2),
            };
        }
        catch (ArgumentException exception) { return Fail(exception.Message, 2); }
        catch (Exception exception) { return Fail(exception.Message, 1); }
    }

    private static int RunConfig(string[] arguments)
    {
        if (arguments.SequenceEqual(["path"]))
        {
            Console.WriteLine(AppPaths.ConfigPath);
            return 0;
        }
        using ConfigStore store = new();
        if (store.LastError is not null) return Fail($"configuration is invalid: {store.LastError}", 1);
        if (arguments.SequenceEqual(["get"]))
        {
            Console.WriteLine(JsonSerializer.Serialize(store.Current, PrettyJson));
            return 0;
        }
        if (arguments.Length != 3 || arguments[0] != "set")
            return Fail("usage: MouseBridge config path|get|set <key> <value>", 2);

        string key = arguments[1];
        string value = arguments[2];
        AppConfig config = store.Current;
        switch (key)
        {
            case "middle": config = config with { MiddleAction = ParseAction(value) }; break;
            case "back": config = config with { BackAction = ParseAction(value) }; break;
            case "forward": config = config with { ForwardAction = ParseAction(value) }; break;
            case "reverse-vertical": config = config with { ReverseVerticalScroll = ParseBool(value) }; break;
            case "reverse-horizontal": config = config with { ReverseHorizontalScroll = ParseBool(value) }; break;
            case "scroll-lines":
                if (!int.TryParse(value, out int lines) || lines is < 0 or > 20) return Fail("scroll-lines must be 0-20", 2);
                config = config with { ScrollLines = lines };
                break;
            case "dpi":
                if (!int.TryParse(value, out int dpi) || dpi is < 400 or > 4000) return Fail("dpi must be 400-4000", 2);
                config = config with { Dpi = dpi };
                break;
            default: return Fail($"unknown configuration key: {key}", 2);
        }
        store.Save(config);
        Console.WriteLine("ok");
        return 0;
    }

    private static string ParseAction(string value)
    {
        if (!Shortcut.TryParse(value, out Shortcut? shortcut)) throw new ArgumentException("invalid shortcut");
        return shortcut!.Canonical;
    }

    private static bool ParseBool(string value) => value.ToLowerInvariant() switch
    {
        "true" => true,
        "false" => false,
        _ => throw new ArgumentException("boolean value must be true or false"),
    };

    private static int Diagnose()
    {
        string version = Assembly.GetExecutingAssembly().GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ?? "unknown";
        RuntimeStatusSnapshot? runtime = RuntimeStatusStore.ReadLive();
        IReadOnlyList<HidInterfaceInfo> candidates = HidDeviceEnumerator.EnumerateM750();
        Console.WriteLine($"version={version}");
        Console.WriteLine($"os={RuntimeInformation.OSDescription}");
        Console.WriteLine($"architecture={RuntimeInformation.ProcessArchitecture}");
        Console.WriteLine($"config={AppPaths.ConfigPath}");
        Console.WriteLine($"log={DiagnosticLog.Shared.FilePath}");
        Console.WriteLine($"mouse-backend={(runtime?.MouseHookRunning == true ? "running" : "not-running")}");
        Console.WriteLine($"mappings-enabled={runtime?.MappingsEnabled.ToString().ToLowerInvariant() ?? "unknown"}");
        Console.WriteLine($"sendinput-failures={runtime?.InjectionFailures ?? 0}");
        Console.WriteLine($"m750-accessible-interfaces={candidates.Count}");
        Console.WriteLine($"hidpp-connected={runtime?.Hidpp.Connected.ToString().ToLowerInvariant() ?? "false"}");
        Console.WriteLine($"hidpp-report-lengths={(runtime?.Hidpp.Connected == true ? $"{runtime.Hidpp.InputReportLength}/{runtime.Hidpp.OutputReportLength}" : "unknown")}");
        Console.WriteLine($"dpi-source={(runtime?.Hidpp.Connected == true ? (runtime.Hidpp.DpisFromDevice ? "device" : "compatibility-fallback") : "unavailable")}");
        Console.WriteLine($"dpi-range={(runtime?.Hidpp.SupportedDpis.Count > 0 ? $"{runtime.Hidpp.SupportedDpis[0]}-{runtime.Hidpp.SupportedDpis[^1]} ({runtime.Hidpp.SupportedDpis.Count})" : "unavailable")}");
        Console.WriteLine($"dpi-current={runtime?.Hidpp.CurrentDpi?.ToString() ?? "unavailable"}");
        Console.WriteLine("uipi-note=SendInput cannot inject into applications running at a higher integrity level");
        return 0;
    }

    private static int Help(int code)
    {
        Console.WriteLine("""
            MouseBridge commands:
              MouseBridge config path
              MouseBridge config get
              MouseBridge config set middle|back|forward <shortcut|none>
              MouseBridge config set reverse-vertical|reverse-horizontal <true|false>
              MouseBridge config set scroll-lines <0-20>
              MouseBridge config set dpi <400-4000>
              MouseBridge diagnose
              MouseBridge --self-test
            """);
        return code;
    }

    private static int Fail(string message, int code)
    {
        Console.Error.WriteLine($"MouseBridge: {message}");
        return code;
    }
}
